docs/walkthroughs/secure-ml-pipeline.md

```diff
@@ -0,0 +1,263 @@
+    ```
+    *Note* If you are running DC/OS Kafka, you can get the `broker` and `port` from running 
+    ```bash
+    dcos <kafka_service> --name=<service_name> endpoints broker
+    ```
+    Likely this will be something like:
+    ```bash
+    dcos kafka --name=secure-kafka endpoints broker
+    ```
+    
+1.  Run the classifier, this will load the model you built and saved to HDFS and use it to classify the messages from the `Spammer`:
+    ```bash
+    dcos spark run --submit-args="\
+    --kerberos-principal=client@LOCAL \
+    --keytab-secret-path=/__dcos_base64___keytab \
+    --keystore-secret-path=__dcos_base64__keystore \
+    --keystore-password=changeit \
+    --private-key-password=changeit \
+    --truststore-secret-path=__dcos_base64__truststore \
+    --truststore-password=changeit \
+    --executor-auth-secret=/spark-auth-secret \
+    --conf spark.cores.max=4 \
+    --conf spark.mesos.uris=http://SMS_DATA.txt,http://JAAS_ARTIFACT.conf \
+    --conf spark.driver.extraJavaOptions=-Djava.security.auth.login.config=/mnt/mesos/sandbox/generic_keytab.jaas \
+    --conf spark.executor.extraJavaOptions=-Djava.security.auth.login.config=/mnt/mesos/sandbox/generic_keytab.jaas \
+    --class SpamHamStreamingClassifier http://DCOS_EXAMPLES.jar top1 <broker>:<port> hdfs:///nb_model true"
+
+    ```   
+    You can look at the `stdout` for this job and see a running average classification accuracy for your spam detector. 
+    
+## Digging deeper
+1.  Training a model with a ML Pipeline. One of the nicest features in later Spark versions is the Pipeline model in Spark-ml. These powerful abstractions allow you to wire together `Transformers` and `Estimators`. See the [Spark ML docs](https://spark.apache.org/docs/latest/ml-pipeline.html) for a thorough explanation. The basic idea is that a `Transformer` will simply computes results from individual datums, whereas `Estimators` need to be _fit_ on all the data.  In our example the `Pipeline` is:
+    ```scala
+    // SpamHam.scala 
+    def spamOrHamPipeline(numFeatures: Int): Pipeline = {
+      val stringIndexer = new StringIndexer()
+        .setInputCol("spamorham")
+        .setOutputCol("label")
+      val tokenizer = new Tokenizer()
+        .setInputCol("message")
+        .setOutputCol("tokens")
+      val hashingTF = new HashingTF()
+        .setInputCol(tokenizer.getOutputCol)
+        .setOutputCol("rawFeatures")
+        .setNumFeatures(numFeatures)
+      val idf = new IDF()
+        .setInputCol(hashingTF.getOutputCol)
+        .setOutputCol("features")
+     val nb = new NaiveBayes()
+     new Pipeline().setStages(Array(stringIndexer, tokenizer, hashingTF, idf, nb))
+    }
+    ```
+    That's it. Here's a description of the various steps:
+    1.  The `StringIndexer` is a `Estimator` that turns our "spam" and "ham" labels into indices (numbers) which are easier for the downstream `Estimators` to understand.
```

Comment:
`StringIndexer` is a `Transformer` to turn categorical values (in this case 'spam', 'ham') into numerical, like (0,1)

Reply:
`extends Estimator` https://spark.apache.org/docs/latest/api/scala/index.html#org.apache.spark.ml.feature.StringIndexer

Reply:
oops. right. :+1: